libraries/AP_Relay/AP_Relay.cpp

```diff
@@ -128,61 +120,237 @@ AP_Relay::AP_Relay(void)
+                             RELAY4_PIN_DEFAULT,
+                             RELAY5_PIN_DEFAULT,
+                             RELAY6_PIN_DEFAULT };
+
+    for (uint8_t i = 0; i < MIN(uint8_t(AP_RELAY_NUM_RELAYS), ARRAY_SIZE(pins)); i++) {
+        // set the default
+        if (pins[i] != -1) {
+            _params[i].pin.set_and_default(pins[i]);
+        }
+    }
+}
+
+void AP_Relay::init()
 {
-    if (instance >= AP_RELAY_NUM_RELAYS) {
-        return;
+    set_defaults();
+
+    convert_params();
+
+    // setup the actual default values of all the pins
+    for (uint8_t instance = 0; instance < AP_RELAY_NUM_RELAYS; instance++) {
+        const int8_t pin = _params[instance].pin;
+        if (pin == -1) {
+            // no valid pin to set it on, skip it
+            continue;
+        }
+
+        const AP_Relay_Params::FUNCTION function = _params[instance].function;
+        if (function <= AP_Relay_Params::FUNCTION::NONE || function >= AP_Relay_Params::FUNCTION::NUM_FUNCTIONS) {
+            // invalid function, skip it
+            continue;
+        }
+
+        if (function == AP_Relay_Params::FUNCTION::RELAY) {
+            // relay by instance number, set the state to match our output
+            const AP_Relay_Params::DEFAULT_STATE default_state = _params[instance].default_state;
+            if ((default_state == AP_Relay_Params::DEFAULT_STATE::OFF) ||
+                (default_state == AP_Relay_Params::DEFAULT_STATE::ON)) {
+
+                set_pin_by_instance(instance, (bool)default_state);
+            }
+        } else {
+            // all functions are supposed to be off by default
+            // this will need revisiting when we support inversion
+            set_pin_by_instance(instance, false);
+        }
     }
-    if (_pin[instance] == -1) {
+}
+
+void AP_Relay::set(const AP_Relay_Params::FUNCTION function, const bool value) {
+    if (function <= AP_Relay_Params::FUNCTION::NONE && function >= AP_Relay_Params::FUNCTION::NUM_FUNCTIONS) {
+        // invalid function
         return;
     }
-    const uint32_t now = AP_HAL::millis();
-    _pin_states = value ? _pin_states | (1U<<instance) : _pin_states & ~(1U<<instance);
-    if ((_pin_states != _last_logged_pin_states) && (now - _last_log_ms > 10)) {
-       AP::logger().Write("RELY", "TimeUS,State", "s-", "F-", "QB",
-                                        AP_HAL::micros64(),
-                                        _pin_states);
-       _last_log_ms = now;
-       _last_logged_pin_states = _pin_states;
+
+    for (uint8_t instance = 0; instance < AP_RELAY_NUM_RELAYS; instance++) {
+        if (function != _params[instance].function) {
+            continue;
+        }
+
+        set_pin_by_instance(instance, value);
+    }
+}
+
+// set a pins output state by instance and log if required
+// this is an internal helper, instance must have already been validated to be in range
+void AP_Relay::set_pin_by_instance(uint8_t instance, bool value) {
```

Comment:
Should be 
```suggestion
void AP_Relay::set_pin_by_instance(uint8_t instance, bool value)
{
```

... the existing code follows this style, MdB (and me, for that matter) prefer the brace on the same line :-)

Reply:
fixed